lxd/storage/drivers/driver_pure_util.go

```diff
@@ -517,6 +572,475 @@ func (p *pureClient) disconnectHostFromVolume(poolName string, volName string, h
 	return nil
 }
 
+// getTarget retrieves the target that can be connected to using the initiator's qualified name.
+func (p *pureClient) getTarget() (*pureTarget, error) {
+	var resp pureResponse[pureTarget]
+
+	targetAddr := p.driver.config["pure.array.address"]
+
+	// Append service port to the address if missing.
+	// This way we can use exact matching in the filter to find the target.
+	if !strings.Contains(targetAddr, ":") {
+		// Default NVMe port is 4420.
+		port := "4420"
+
+		if p.driver.config["pure.mode"] == pureModeISCSI {
+			// Default iSCSI port is 3260.
+			port = "3260"
+		}
+
+		targetAddr = fmt.Sprintf("%s:%s", targetAddr, port)
+	}
+
+	err := p.requestAuthenticated(http.MethodGet, fmt.Sprintf("/ports?filter=portal='%s'", targetAddr), nil, &resp)
+	if err != nil {
+		return nil, fmt.Errorf("Failed to retrieve PureStorage targets for address %q: %w", targetAddr, err)
+	}
+
+	if len(resp.Items) == 0 {
+		return nil, api.StatusErrorf(http.StatusNotFound, "No PureStorage target found on address %q", targetAddr)
+	}
+
+	return &resp.Items[0], nil
+}
+
+// ensureHost returns a name of the host that is configured with a given IQN. If such host
+// does not exist, a new one is created, where host's name equals to the server name with a
+// mode included as a suffix because PureStorage does not allow mixing IQNs, NQNs, and WWNs
+// on a single host.
+func (d *pure) ensureHost() (hostName string, cleanup revert.Hook, err error) {
+	var hostname string
+
+	revert := revert.New()
+	defer revert.Fail()
+
+	switch d.config["pure.mode"] {
+	case pureModeISCSI:
+		iqn, err := d.hostIQN()
+		if err != nil {
+			return "", nil, err
+		}
+
+		// Fetch the host by IQN.
+		host, err := d.client().getHostByIQN(iqn)
+		if err != nil {
+			if !api.StatusErrorCheck(err, http.StatusNotFound) {
+				return "", nil, err
+			}
+
+			// The host with a given IQN does not exist, therefore, create a new one.
+			hostname, err = d.serverName()
```

Comment:
is the assumption that if the LXD host becomes clustered after the pool is setup the then IQN mapping will break?

Reply:
Huh, a very good point. But it should not break. The host's name will be different and it will result in each LXD cluster member having its own host with different IQN (that is read from the `/etc/iscsi/initiatorname.iscsi`). Individual host will then be mapped with the volumes that are mounted on that particular host.